aggregator/src/aggregator/queue.rs

```diff
@@ -0,0 +1,567 @@
+            }
+        })
+    }
+
+    fn acquire(&self) -> Result<Ticket, Error> {
+        // Unwrap safety: mutex poisoning.
+        let mut stack = self.stack.lock().unwrap();
+        if stack.len() < self.depth {
+            let (permit_tx, permit_rx) = oneshot::channel();
+            let id = self.id_generator.next();
+            let dispatcher = Dispatcher {
+                id,
+                permit_tx: Some(permit_tx),
+            };
+            let ticket = Ticket {
+                id,
+                permit_rx: Some(permit_rx),
+                cancel_tx: self.cancel_tx.clone(),
+            };
+
+            stack.push(dispatcher);
+            self.notifier.notify_waiters();
+            Ok(ticket)
+        } else {
+            Err(Error::TooManyRequests)
+        }
+    }
+
+    #[cfg(test)]
+    pub fn len(&self) -> usize {
+        // Unwrap safety: mutex poisoning.
+        let stack = self.stack.lock().unwrap();
+        stack.len()
+    }
+
+    #[cfg(test)]
+    pub fn available_permits(&self) -> usize {
+        self._semaphore.available_permits()
+    }
+}
+
+impl Drop for LIFORequestQueue {
+    fn drop(&mut self) {
+        self.dispatcher.abort();
+        self.canceller.abort();
+    }
+}
+
+/// Identifies a ticket in the queue.
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+struct TicketId(usize);
+
+/// Simple generator for unique ticket numbers. Ticket numbers are unique as long as there aren't
+/// more than [`usize`] entries in the queue, at which point you've run out of memory.
+#[derive(Debug, Default)]
+struct TicketIdGenerator(AtomicUsize);
+
+impl TicketIdGenerator {
+    fn next(&self) -> TicketId {
+        TicketId(self.0.fetch_add(1, Ordering::Relaxed))
+    }
+}
+
+/// Dispatcher-side ticket which is used to signal when the ticket is ready. Corresponds 1:1 with a
+/// [`Ticket`].
+#[derive(Debug)]
+struct Dispatcher {
+    /// Identifies the ticket in the queue.
+    id: TicketId,
+
+    /// Signals the [`Ticket`] holder that it may proceed.
+    permit_tx: Option<oneshot::Sender<OwnedSemaphorePermit>>,
+}
+
+impl Dispatcher {
+    fn send(&mut self, permit: OwnedSemaphorePermit) {
+        // Unwrap safety: we should only send once on this channel.
+        let _ = self
+            .permit_tx
+            .take()
+            .unwrap()
+            .send(permit)
+            .map_err(|_| warn!("failed to dispatch, request cancelled?"));
+    }
+}
+
+/// Handler-side ticket which is used to get in the queue.
+#[derive(Debug)]
+struct Ticket {
+    id: TicketId,
+    permit_rx: Option<oneshot::Receiver<OwnedSemaphorePermit>>,
+    cancel_tx: mpsc::UnboundedSender<TicketId>,
+}
+
+impl Ticket {
+    /// Waits for the ticket to be called. Returns an [`OwnedSemaphorePermit`] which should be
+    /// dropped to signal that the request is done.
+    async fn wait(&mut self) -> Result<OwnedSemaphorePermit, Error> {
+        // Unwrap safety: wait() should only be called once.
```

Comment:
Q. Could we guarantee this by making `wait` take `self` rather than `&mut self`? This would let us drop the `Option` around `permit_rx`, too, I think.

Reply:
I think not really. We can't implement Drop and have a method that takes `self` at the same time. We could move `Drop` code to a manually invoked `.cancel()` method, but then we run into use-after-move problems in `run()`.